File: nodejs/src/api/handlers/forms/getFormByTitle/getFormByTitleHandler.js
'use strict';

const AWS = require('aws-sdk')
const db = new AWS.DynamoDB.DocumentClient({})

const tableName = process.env.FORMS_TABLE_NAME

const response = (statusCode, message) => {
    return {
        statusCode: statusCode,
        body: JSON.stringify(message)
    }
}

const getFormByTitle = async (event) => {
    const title = event.queryStringParameters.title
    console.log('title: ', title)
  
    const params = {
      TableName: tableName,
      IndexName: 'SearchFormByTitle',
      KeyConditionExpression: 'title = :v AND begins_with(PK, :v1)',
      ExpressionAttributeValues: {
        ':v': title,
        ':v1': 'FORM'
      }
    }
  
    try {
      const formSearch = await db.query(params).promise()
      console.log('formSearch', formSearch)
  
      if (!formSearch.Items || formSearch.Items.length !== 1) {
        return response(500, {
          message: "Form not found"
        })
      }
  
      const form = formSearch.Items[0]
      
      return response(200, form)
    } catch (err) {
      console.log('err: ', err)
      return response(err.statusCode, err)
    }
}

module.exports = {
    getFormByTitle
}

File: python/src/api/handlers/forms/getFormByTitle/getFormByTitleHandler.py
try:
    import unzip_requirements
except ImportError:
    pass

import boto3
import os
import simplejson as json

dynamodb_client = boto3.resource('dynamodb')
TABLE_NAME = os.environ['FORMS_TABLE_NAME']
DYNAMO_FORM_ID = os.environ['FORMS_ID']

def create_response(code, message):
  return {
    'statusCode': code,
    'headers': {'Content-Type': 'application/json'},
    'body': json.dumps(message)
  }

def getFormByTitle(event, context):
    title = event['queryStringParameters']['title']

    params = {
        'IndexName': 'SearchFormByTitle',
        'KeyConditionExpression': 'title = :v AND begins_with(PK, :v1)',
        'ExpressionAttributeValues': {
            ':v': title,
            ':v1': 'FORM'
        }
    }

    try:
        table = dynamodb_client.Table(TABLE_NAME)
        response = table.query(**params)
    except Exception as ex:
        return create_response(500, str(ex))

    print('response: ', response)

    if response['ResponseMetadata']['HTTPStatusCode'] != 200:
      return create_response(500, 'Error getting the Form')

    if not response['Items'] or len(response['Items']) != 1:
        return create_response(500, { 'message': 'Form not found' })

    form = response['Items'][0]

    return create_response(200, form)

